posthog/hogql/printer.py

```diff
@@ -561,7 +598,132 @@ def visit_lambda(self, node: ast.Lambda):
+                    # If we're comparing to an empty string literal, we need to disambiguate this from the default value
+                    # for the ``Map(String, String)`` type used for storing property group values by also ensuring that
+                    # the property key is present in the map. If this is in a ``WHERE`` clause, this also ensures we can
+                    # still use the data skipping index on keys, even though the values index cannot be used.
+                    printed_expr = f"and({property_source.has_expr}, {printed_expr})"
+
+                return printed_expr
+
+            elif node.op == ast.CompareOperationOp.NotEq:
+                if constant_expr.value is None:
+                    # "IS NOT NULL" can be interpreted as "does exist in the map" -- this avoids unnecessarily reading
+                    # the ``values`` subcolumn of the map, and also allows us to use the data skipping index on keys.
+                    return property_source.has_expr
+
+        elif node.op in (ast.CompareOperationOp.In):
+            # ``IN`` is _not_ commutative, so we only need to check the left side operand (in contrast with above.)
+            left_type = resolve_field_type(node.left)
+            if not isinstance(left_type, ast.PropertyType):
+                return None
+
+            # TODO: Chained properties could likely be supported here to at least use the keys index.
+            if left_type is None or len(left_type.chain) > 1:
+                return None
+
+            property_source = self.__get_materialized_property_source(left_type)
+            if not isinstance(property_source, PrintableMaterializedPropertyGroupItem):
+                return None
+
+            if isinstance(node.right, ast.Constant):
+                if node.right.value is None:
+                    return "0"
+                elif node.right.value == "":
+                    # If the RHS is the empty string, we need to disambiguate it from the default value for missing keys.
+                    return f"and({property_source.has_expr}, equals({property_source.value_expr}, {self.visit(node.right)}))"
+                else:
+                    return f"in({property_source.value_expr}, {self.visit(node.right)})"
+            elif isinstance(node.right, ast.Tuple):
+                # If any of the values on the RHS are the empty string, we need to disambiguate it from the default
+                # value for missing keys. NULLs should also be dropped, but everything else can be passed through as-is.
+                default_value_expr: ast.Constant | None = None
+                for expr in node.right.exprs[:]:
+                    if not isinstance(expr, ast.Constant):
+                        return None  # only optimize constants for now, see above
+                    elif expr.value is None:
+                        node.right.exprs.remove(expr)  # XXX: safe to mutate?
```

Comment:
It should be relatively safe. Whatever ends up in the printer has most likely been `clone_expr`-d at one of the previous steps.